### ISSUE TITLE
Override database URL at runtime

### ISSUE THREAD

**Issue by jakeleventhal** [prisma/prisma#18314]
## Problem

I am trying to migrate my repository to use bazel. I want to be able to have hermetic tests for this that allow me to not have to have a giant global db running in the background. Each test should be able to run independently and in parallel without worrying about DB collisions (this is a problem without bazel too).

## Suggested solution

I want to accomplish this with a tool like [testcontainers](https://github.com/testcontainers/testcontainers-node). However, in prisma the database URL is specified in the schema.prisma. If I could override this at runtime, each test could override the database URL to a dynamically generated one for each test and allow for parallel tests to run.

**Comment by janpio** [comment]
Good news then: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#programmatically-override-a-datasource-url

**Comment by jakeleventhal** [comment]
Spoke offline, we should reopen this issue.

The problem is that if you want parallel tests, it would be nice to not have to mock the client. Here is my current implementation to setup a test. This feature would allow me to skip the steps on lines 35-47 in my screenshot.
<img width="1204" alt="Screenshot 2023-03-17 at 9 07 04 AM" src="https://user-images.githubusercontent.com/15133828/225913209-c9b8cab9-6866-4eb0-a8e1-99c4818f5a2c.png">

In production, I export the prisma client (instantiated) from my `@ecominate/database` package. I wouldn't have to any client mocking if there were some setter method available for the database URL like so:
```typescript
// import prisma (this is an instantiated PrismaClient using env("DATABASE_URL") which points to a local dev db)
import prisma from "@ecominate/database"
 
// below would be in a unit test
prisma.setDatabaseUrl(<url goes here>);
 
// everything from now on accesses the new url which is a db url that exists only for the duration of the test
```

**Comment by janpio** [comment]
I don't get the connection between the need for mocking, and that you currently can only change the connection string when creating a Prisma Client. 

Let me try to explain:
If your instance of `prismaClient`is only used in one test, you will get exactly the same result. If it is used in multiple tests, that can not work as calling `setDatabaseUrl` would modify the Prisma Client instance that currently might be in use in parallel in other tests. What am I missing?

Besides that, being able to set a new connection string for the database and letting Prisma "restart" with that sounds like a good API to add.

**Comment by jakeleventhal** [comment]
@janpio the problem is actually fine

**Comment by jakeleventhal** [comment]
The problem is actually resolved

Maybe this warrants a separate issue, but it's somewhat related to runtime override:
```ts
const client = (datasourceUrl: string | undefined) => new PrismaClient({ datasourceUrl });

// Export the prisma client
export const getPrismaClient = (url?: string): PrismaClient => {
  // @ts-ignore
  if (typeof window === 'undefined') {
    return client(url);
  }

  // @ts-ignore
  return null;
};

const prisma = getPrismaClient();
export default prisma;
export * from '@prisma/client;
```
I do this in a shared pnpm workspace package so that if i use this package in the backend, i can import the prisma client just fine. But if i want to import a type in the frontend, like a User model, it doesn't try to connect to the DB in the browser. Is there a better way to do this? Shutting of DB connection attempts via some runtime override would simplify this code and make it such that i dont have to do `// @ts-ignore` to lie to TS

**Comment by janpio** [comment]
Definitely new issue please, then I can pull in the right people. Thanks.

And this one can be closed?

**Comment by jakeleventhal** [comment]
Yes agreed. Here is the new issue I just created: https://github.com/prisma/prisma/issues/21126
